### ISSUE TITLE
Stability due to action scaling and discrete control support

### ISSUE THREAD

**Issue by oshapio** [lmzintgraf/varibad#20]
Hi,
thanks for open-sourcing the code and making it easy to run! 

I have a few issues that prevent from utilizing your method well (and I wouldn't want to underrepresent it) and would appreciate some pointers:

1. **Instability in PPO when no action bounding is used**.

This was rather surprising since training vanilla PPO on the underlying environment without action bounding never posed challenges. Adding action squashing (`norm_actions_post_sampling=True`) leads to a stable training, but the performance becomes no longer on par with the non-normed version before the divergence happens (and the sample efficiency is way lower). I've tried VariBAD on two continuous control envs (though not the ones from your repo) with a similar behavior.

Lowering further the learning rate only postpones the issues to a later epoch, so I'd guess the issue is not there.

What seems to happen concretely is this: when doing PPO update on epoch >= 1, the log ratio takes exceedingly _low_ values, so PPO clipping isn't activated (i.e. some action which should be encouraged, likely gets smaller density due to another action in the collected data). This happens because for some reason the actor shifts its mean to a completely different location, and so the ratio becomes very high for advantage that is negative, and this doesn't get clipped. I was only able to ~fix~ improve this issue after setting higher $\epsilon$ (from 1e-8 to 1e-5) in the Adam optimizer.

Have you encountered something like this? I see that in your configs you don't use pre/post actions norm, but in some HyperX configs you use normalization (e.g. [1]). In general, is there some intuition when you suggest to use the action scaling? In some cases, the mean of the action seems to grow indefinitely, so it's probably something I need to take care of.

2. **Using action embedding in discrete control**.

This seems to not be supported currently, right? I guess if we store indices of the actions in the VAE's storage, and use standard embedders (e.g. torch.nn.Embedding) in the encoder and the decoder, it should be capable of accounting for discrete actions too?

Thanks!

[1] https://github.com/lmzintgraf/hyperx/blob/main/config/sparse_ant_goal/args_sparse_ant_goal_varibad.py

**Comment by lmzintgraf** [comment]
Hi, thanks for your questions! 

I never looked very closely into the effects of action scaling; usually it was fine to keep it turned off but in some cases it helped with some learning instabilities (e.g. in Humaoid iirc). A good strategy for me was to tune hyperparameters on a single-agent version first, and those would often transfer well to the meta-learning setting. 

The code doesn't currently support discrete control action embeddings, they're just passed to the network as indices or one-hots. This worked well in the gridworld, but I think there the network can probably just memorise most things. Using standard embedders should probably work :)
